ramips: add support for TP-Link Archer C50 v6 (CA/EU/RU)

jschwartzenberg opened the pull request:
Related to the forum discussion: https://forum.openwrt.org/t/tp-link-archer-c50-eu-v6-0-5-ghz-interface-not-present/95532

It would be great if this could be integrated!

Renaud11232 committed:
ramips: Add support for TP-Link Archer C50 v6 (CA/EU/RU)

This adds support for the TP-Link Archer C50 v6 (CA/EU/RU).
(The ES variant is a rebranded Archer C54 and NOT supported.)

CPU:   MediaTek MT7628 (580MHz)
RAM:   64M DDR2
FLASH: 8M SPI
WiFi:  2.4GHz 2x2 MT7628 b/g/n integrated
WiFi:  5GHz 2x2 MT7613 a/n/ac
ETH:   1x WAN 4x LAN
LED:   Power, WiFi2, WiFi5, LAN, WAN, WPS
BTN:   WPS/WiFi, RESET
UART:  Near ETH ports, 115200 8n1, TP-Link pinout

Create Factory image
--------------------
As all installation methods require a U-Boot to be integrated into the
image (and we do not ship one with the image). We are not able to create
an image in the OpenWRT build-process.

Download a TP-Link image for your device variant (CA/EU or RU) from their
website and a OpenWRT sysupgrade image for the device
and build yourself a factory image like following:

TP-Link image:             tpl.bin
OpenWRT sysupgrade image:  owrt.bin

 > dd if=tpl.bin of=boot.bin bs=131584 count=1
 > cat owrt.bin >> boot.bin

Installing via Web-UI
---------------------
Upload the boot.bin via TP-Links firmware upgrade tool in the
web-interface.

Installing via Recovery
-----------------------
Activate Web-Recovery by beginning the upgrade Process with a
Firmware-Image from TP-Link. After starting the Firmware Upgrade,
wait ~3 seconds (When update status is switching to 0%), then
disconnect the power supply from the device. Upgrade flag (which
activates Web-Recovery) is written before the OS-image is touched and
removed after write is succesfull, so this procedure should be safe.

Plug the power back in. It will come up in Recovery-Mode on 192.168.0.1.
When active, all LEDs but the WPS LED are off.
Remeber to assign yourself a static IP-address as DHCP is not active in
this mode.

The boot.bin can now be uploaded and flashed using the web-recovery.

Installing via TFTP
-------------------
Prepare an image like following (Filenames from factory image steps
apply here)

 > dd if=/dev/zero of=tp_recovery.bin bs=196608 count=1
 > dd if=tpl.bin of=tmp.bin bs=131584 count=1
 > dd if=tmp.bin of=boot.bin bs=512 skip=1
 > cat boot.bin >> tp_recovery.bin
 > cat owrt.bin >> tp_recovery.bin

Place tp_recovery.bin in root directory of TFTP server and listen on
192.168.0.66/24.

Connect router LAN ports with your computer and power up the router
while pressing the reset button. The router will download the image via
tftp and after ~1 Minute reboot into OpenWRT.

U-Boot CLI
----------
U-Boot CLI can be activated by holding down '4' on bootup.

Dual U-Boot
-----------
This is the first TP-Link MediaTek device to feature a split-uboot
design. The first (factory-uboot) provides recovery via TFTP and HTTP,
jumping straight into the second (firmware-uboot) if no recovery needs
to be performed. The firmware-uboot unpacks and executed the kernel.

Web-Recovery
------------
TP-Link integrated a new Web-Recovery like the one on the Archer C7v4 /
TL-WR1043v5. Stock-firmware sets a flag in the "romfile" partition
before beginning to write and removes it afterwards. If the router boots
with this flag set, bootloader will automatically start Web-recovery and
listens on 192.168.0.1. This way, the vendor-firmware or an OpenWRT
factory image can be written.

By doing the same while performing sysupgrade, we can take advantage of
the Web-recovery in OpenWRT.

It is important to note that Web-Recovery is only based on this flag. It
can't detect e.g. a crashing kernel or other means. Once activated it
won't boot the OS before a recovery action (either via TFTP or HTTP) is
performed. This recovery-mode is indicated by an illuminated WPS-LED on
boot.

Co-authored-by: Julius Schwartzenberg <julius.schwartzenberg@gmail.com>
Signed-off-by: Renaud Gaspard <gaspardrenaud@hotmail.com>
Signed-off-by: Julius Schwartzenberg <julius.schwartzenberg@gmail.com>
Tested-by: Julius Schwartzenberg <julius.schwartzenberg@gmail.com>
Tested-by: Jaroslav Mikulík <byczech@gmail.com>
Tested-by: Ashipa Eko <ashipa.eko@gmail.com>

GoliathLabs commented:
There's already a PR for the C50 v6, would be nice if you could rename the PR to "ramips: add support for TP-Link Archer C50 v6".

https://github.com/openwrt/openwrt/pull/5046

jschwartzenberg commented:
> There's already a PR for the C50 v6, would be nice if you could rename the PR to "ramips: add support for TP-Link Archer C50 v6".
> 
> #5046

My apologies, I had missed that.


@Renaud11232 as I took your patch, the checks are looking for your name as a `Signed-off-by`. Are you ok with me adding this header for your name?

herbetom commented:
@jschwartzenberg I think how it works is that if you did changes best practice is to note that and then sign it off yourself. I'm not sure what's the threshold to also add Co-authored-by.

An example to see this in action might be: 7efec0acca80b231ab8e69729a4bdaf11ef84541 (first result, YMMV)

https://openwrt.org/submitting-patches#sign_your_work

jschwartzenberg commented:
> @jschwartzenberg I think how it works is that if you did changes best practice is to note that and then sign it off yourself. I'm not sure what's the threshold to also add Co-authored-by.

Thanks for the info! At this point I made zero changes to this, I'm just submitting the original patch as-is, just rebased against the latest main branch. If changes are needed I will make them and introduce `Co-authored-by`.

herbetom commented:
@jschwartzenberg I do think the threshold for adding your sign off is that you're the one submitting it. The reason is that you're thereby certifying that you've got the right to submit it.

That you've rebased it should also be indicated like it's shown in the Wiki or done in the Example. There's alwas a possibilty to mess up during a rebase and that way it's clear you're taking over that responsibility. From my understanding this is hoewer more good practice and not a requirement like the sign off.

jschwartzenberg commented:
> @jschwartzenberg I do think the threshold for adding your sign off is that you're the one submitting it. The reason is that you're thereby certifying that you've got the right to submit it.

I added a sign off for myself as well now.


> That you've rebased it should also be indicated like it's shown in the Wiki or done in the Example. There's alwas a possibilty to mess up during a rebase and that way it's clear you're taking over that responsibility. From my understanding this is hoewer more good practice and not a requirement like the sign off.

I don't know if there is a special way to indicate that I did a rebase? I should note that I did not need to do anything "by hand" for the rebase. I merged the commits from @Renaud11232's branch into a single commit (fixup) and changed the message. Since then I blindly did `git rebase upstream/main` and everything always applied cleanly without any need for manual adjustment.

Just now I did add a small extra note to the commit message. This supports the EU and RU models, but not the ES model.

jschwartzenberg commented:
It seems other models specify the region explicitly when applicable:
![image](https://github.com/openwrt/openwrt/assets/877112/4b89fac1-54b7-4d46-b811-aeb1f145d811)

I'll update the commit to match what was done for those models.

GoliathLabs commented:
@jschwartzenberg can you rebase as one workflow failed

ByCzech reviewed (approved):
Tested on newly buyed router v6 EU for few hours with many devices on ethernet, WiFi 2.4 GHz and 5 GHz too. Working like a charm with no problems, no hungs, no restart needed.

jschwartzenberg commented:
Added @ByCzech as a tester and remove the known issue as it seems this cannot be reproduced anymore. I suspect it was resolved by updates in OpenWRT (and rebasing).

rallep71 commented:
Is there anywhere to download the firmware?

thx

jschwartzenberg commented:
> Is there anywhere to download the firmware?

The ODM firmware can be downloaded from TP-Link's site.

For OpenWRT I manually did some builds that can be found in the related forum topic: https://forum.openwrt.org/t/tp-link-archer-c50-eu-v6-0-5-ghz-interface-not-present/95532

Note that these builds still need the bootloader prepended to be able to upgrade from the ODM firmware.

rallep71 commented:
> > Is there anywhere to download the firmware?
> 
> The ODM firmware can be downloaded from TP-Link's site.
> 
> For OpenWRT I manually did some builds that can be found in the related forum topic: https://forum.openwrt.org/t/tp-link-archer-c50-eu-v6-0-5-ghz-interface-not-present/95532
> 
> Note that these builds still need the bootloader prepended to be able to upgrade from the ODM firmware.
@jschwartzenberg
Could you post the instructions on how to do this exactly with the bootloader and the original file, something with cat >>>? 

thx

jschwartzenberg commented:
> Could you post the instructions on how to do this exactly with the bootloader and the original file, something with cat >>>?

Please take a look at the commit message of the commit related to the PR. You can find the commits tab at the top. I copied the message mostly from the commit adding v4 support only correcting some typos and adding a note on the variants.
If you have any feedback on how the instructions could be improved, that might be useful to incorporate. The instructions there should be complete enough to get it to work.

rallep71 commented:
After install i wont to update and the follow errors pop up

Executing package manager
Upgrading base-files on root from 1542-r24122+32-4e5d45f1e6 to 1542-r24154-6f31941d38...
Downloading https://downloads.openwrt.org/snapshots/targets/ramips/mt76x8/packages/base-files_1542-r24154-6f31941d38_mipsel_24kc.ipk
Errors
umount: tmpfs busy - remounted read-only
umount: can't remount tmpfs read-only
umount: proc busy - remounted read-only
Collected errors:
 * copy_file: unable to open `/etc/group-opkg.backup': Read-only file system.
 * file_copy: Failed to copy file /etc/group to /etc/group-opkg.backup.
 * backup_make_backup: Failed to copy /etc/group to /etc/group-opkg.backup
 * opkg_install_cmd: Cannot install package base-files.
 * pkg_write_filelist: Failed to open //usr/lib/opkg/info/base-files.list: Read-only file system.
The opkg install command failed with code 255.



Executing package manager
Upgrading base-files on root from 1542-r24122+32-4e5d45f1e6 to 1542-r24154-6f31941d38...
Downloading https://downloads.openwrt.org/snapshots/targets/ramips/mt76x8/packages/base-files_1542-r24154-6f31941d38_mipsel_24kc.ipk
Errors
Command failed: ubus call service delete { "name": "system" } (Not found)
Command failed: ubus call service delete { "name": "gpio_switch" } (Not found)
umount: tmpfs busy - remounted read-only
umount: can't remount tmpfs read-only
umount: proc busy - remounted read-only
Collected errors:
 * copy_file: unable to open `/etc/group-opkg.backup': Read-only file system.
 * file_copy: Failed to copy file /etc/group to /etc/group-opkg.backup.
 * backup_make_backup: Failed to copy /etc/group to /etc/group-opkg.backup
 * opkg_install_cmd: Cannot install package base-files.
 * pkg_write_filelist: Failed to open //usr/lib/opkg/info/base-files.list: Read-only file system.
The opkg install command failed with code 255.

jschwartzenberg commented:
> After install i wont to update and the follow errors pop up

I don't think you can use `opkg` like that with self-built images. You'll have to build any needed packages as part of your own build.

hauke commented:
I just wanted to merge this and then saw that in f53fa2a0cb9c3c802ac79409330bcba3b971c8ca the EEPROM handling was updated for the other similar devices. Could you please do the same. 
I think this device has a bigger eeprom of 0x4da8 bytes. You probably have to overwrite the eeprom_radio_8000 entry too.

jschwartzenberg commented:
> I just wanted to merge this and then saw that in [f53fa2a](https://github.com/openwrt/openwrt/commit/f53fa2a0cb9c3c802ac79409330bcba3b971c8ca) the EEPROM handling was updated for the other similar devices. Could you please do the same.

I am not fully understanding and I'm just copying the v4 change, this gives me this diff:
```
diff --git a/target/linux/ramips/dts/mt7628an_tplink_archer-c50-v6.dts b/target/linux/ramips/dts/mt7628an_tplink_archer-c50-v6.dts
index 4342aa61b5..fb20be6dfa 100644
--- a/target/linux/ramips/dts/mt7628an_tplink_archer-c50-v6.dts
+++ b/target/linux/ramips/dts/mt7628an_tplink_archer-c50-v6.dts
@@ -87,10 +87,9 @@
 &pcie0 {
        wifi@0,0 {
                reg = <0x0000 0 0 0 0>;
-               mediatek,mtd-eeprom = <&radio 0x8000>;
                ieee80211-freq-limit = <5000000 6000000>;
-               nvmem-cells = <&macaddr_rom_f100>;
-               nvmem-cell-names = "mac-address";
+               nvmem-cells = <&eeprom_radio_8000>, <&macaddr_rom_f100>;
+               nvmem-cell-names = "eeprom", "mac-address";
                mac-address-increment = <(-1)>;
        };
 };

```
Does that reflect the (first) change?


> I think this device has a bigger eeprom of 0x4da8 bytes. You probably have to overwrite the eeprom_radio_8000 entry too.

That means overriding the entry defined in `mt7628an_tplink_8m-split-uboot.dtsi` in the device specific dts? The C50 v4 variant does not seem to have this section overridden in its dts. Does the bigger eeprom of 0x4da8 bytes relate to the 5 GHz WiFi chip or is there another reason you think the v6 has the bigger eeprom? Is there a mechanism to find out/verify the size of the eeprom?

Thanks for checking this out! I'll try to accommodate any changes necessary. I can't test myself in the upcoming weeks (I'm in another country than my C50 now), hopefully others on the forum can test in the mean time.

hauke commented:
The eeprom integration was changed in f53fa2a0cb9c3c802ac79409330bcba3b971c8ca by @DragonBluep .
The commit comment says this:
> there are only three devices uses
> MT7613 NIC, ASUS RT-AC1200 V2, COMFAST CF-WR758AC V2 and Keenetic
> KN-1613. The EEPROM size of them is 0x4da8.

I think you have to add this into your dts file in addition to the change you already suggested in the comment:
````
&eeprom_radio_8000 {
	reg = <0x8000 0x4da8>;
};
````

This will overwrite the reg property in the `eeprom_radio_8000` node in your dts file.

jschwartzenberg commented:
The needed diff is clear, thanks for making it explicit, I saw similar additions to those three devices DTS files indeed.

But would the Archer C50 v6 indeed be the forth device that uses the different eeprom size? Because of the MT7613?

jschwartzenberg commented:
Ah yeah if it's about all the devices with the MT7613 then it seems likely to apply for this one too! Great you are paying close attention!!

I pushed the changes just now.

GoliathLabs commented:
@rallep71 can you try out the new build so we can get it merged?

rallep71 commented:
> @rallep71 can you try out the new build so we can get it merged?

@GoliathLabs I would like to do this if I had the new build, where can I download it?

jschwartzenberg commented:
> Is this now a build where you can also install your own software or as described above, i.e. everything is hard-coded?

It's the same as prior builds. (I think regular installing can only be done after this merge and a new OpenWRT release.)

GoliathLabs commented:
@jschwartzenberg can you build the new build so it can be tested properly? Would be nice <3

jschwartzenberg commented:
> @jschwartzenberg can you build the new build so it can be tested properly? Would be nice <3

The build linked above resembles the current state of this PR. Should I do both a rebase & rebuild?

Renaud11232 commented:
Hello,

I've managed to find some time to install the snapshot on my router and so far it's been running flawlessly for 4 days.
The 5GHz radio is capable of delivering at least 27Mb/s (which is my full WAN bandwidth) on channel 36 with 2-3 devices connected.
Nothing worying shows up in the kernel or system logs in LuCI.

GoliathLabs commented:
So this PR can be merged then, right? @hauke

DragonBluep commented on target/linux/ramips/dts/mt7628an_tplink_archer-c50-v6.dts:
```diff
@@ -0,0 +1,99 @@
+};
+
+&eeprom_radio_8000 {
+	reg = <0x8000 0x4da8>;
```
Adding some comments here should be better. Just like asus rt-ac1200 v2.

jschwartzenberg replied:
Copied the comment from `mt7628an_asus_rt-ac1200-v2.dts`! Thanks!

DragonBluep commented on target/linux/ramips/image/mt76x8.mk:
```diff
@@ -593,6 +593,20 @@ define Device/tplink_archer-c50-v4
+  TPLINK_HWREVADD := 0x6
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7663-firmware-ap
+  IMAGES := sysupgrade.bin
+  SUPPORTED_DEVICES += tplink,c50-v6
```
`SUPPORTED_DEVICES` can be removed as it's a newly added device.

jschwartzenberg replied:
I removed this line.

jschwartzenberg commented:
[openwrt-ramips-mt76x8-tplink_archer-c50-v6-squashfs-sysupgrade.bin.gz](https://github.com/openwrt/openwrt/files/13324912/openwrt-ramips-mt76x8-tplink_archer-c50-v6-squashfs-sysupgrade.bin.gz)

Renaud11232 commented:
The router crashed today. It became unresponsive, all leds were off.
I'll try to hook a serial adapter to it to check if that gives useful info when/if it happens again

dovanhuong commented:
> [openwrt-ramips-mt76x8-tplink_archer-c50-v6-squashfs-sysupgrade.bin.gz](https://github.com/openwrt/openwrt/files/13324912/openwrt-ramips-mt76x8-tplink_archer-c50-v6-squashfs-sysupgrade.bin.gz)

hi expert, when I ported this firmware into my TP-Link archer C50 V6 it did not run with LED and all the LED is off, could you please help to check in my case I used a router of tp-link with flash XMC instead of Windbond.

dovanhuong commented:
thanks, @jschwartzenberg, I also tried the binary image built from the link you shared, unfortunately, it seems not to be working.

th1nhhdk commented:
Can anyone give me a status update about this PR? I can test but I don't know how to unbrick it when it fails (I'm a newbie)

jschwartzenberg commented:
> Can anyone give me a status update about this PR? I can test but I don't know how to unbrick it when it fails (I'm a newbie)

All information should be on the wiki page for this device. Did you read through it? Is any information missing?

th1nhhdk commented:
> All information should be on the wiki page for this device. Did you read through it? Is any information missing?

Last time I checked it said that v6 have a PR but not merged yet, now it's updated

But this part (https://openwrt.org/toh/tp-link/archer-c50#notes_and_known_issues) is outdated and there are no installation instructions

I managed to get it working but it's hard without any information in the wiki

And my router crashed 1 time when changing Wi-fi settings, have to reboot but no crashes since then

jschwartzenberg commented:
> But this part (https://openwrt.org/toh/tp-link/archer-c50#notes_and_known_issues) is outdated and there are no installation instructions

I changed these parts, pointing to the installation instructions of the commit message. Would those be clear enough? It was indeed not explicit the instructions were basically matching the v4 ones.

th1nhhdk commented:
> > But this part (https://openwrt.org/toh/tp-link/archer-c50#notes_and_known_issues) is outdated and there are no installation instructions
> 
> I changed these parts, pointing to the installation instructions of the commit message. Would those be clear enough? It was indeed not explicit the instructions were basically matching the v4 ones.

Its good enough, thanks

ozgunawesome commented:
can I backport this to 23.05 or is there already an effort going on?

jschwartzenberg commented:
> can I backport this to 23.05 or is there already an effort going on?

I don't think anybody is looking into this. A backport to 23.05 would seem great!

johnsonjamees commented:
On the OpenWRT firmware selector page, version `24.10.0-rc2 (r28161-ea17e958b9)` is available for the **TP-Link Archer C50 v6 (CA/EU/RU)**. Do the installation instructions in the commit message still apply? More precisely, does it require a U-Boot to be integrated into the OpenWRT `SYSUPGRADE` image? For the first-time installation, should I use the `KERNEL` image or the `SYSUPGRADE` image, as stated in the instructions?

jschwartzenberg commented:
> On the OpenWRT firmware selector page, version `24.10.0-rc2 (r28161-ea17e958b9)` is available for the **TP-Link Archer C50 v6 (CA/EU/RU)**. Do the installation instructions in the commit message still apply? More precisely, does it require a U-Boot to be integrated into the OpenWRT `SYSUPGRADE` image? For the first-time installation, should I use the `KERNEL` image or the `SYSUPGRADE` image, as stated in the instructions?

Nothing has changed as far as I know, also presuming newer ODM firmware hasn't significantly changed. I don't think the `kernel` image is relevant for this model. For the first time installation from ODM firmware, you need to follow the instructions as described to prepend the ODM bootloader correctly to the `sysupgrade` image. Once you're on OpenWRT you can use the `sysupgrade` image as usual.

I did not test this version or any recent OpenWRT versions, so I'm curious about people's experiences with 24.10-rc2 on this device!

martamusikmaschine commented:
> On the OpenWRT firmware selector page, version `24.10.0-rc2 (r28161-ea17e958b9)` is available for the **TP-Link Archer C50 v6 (CA/EU/RU)**. Do the installation instructions in the commit message still apply? More precisely, does it require a U-Boot to be integrated into the OpenWRT `SYSUPGRADE` image? For the first-time installation, should I use the `KERNEL` image or the `SYSUPGRADE` image, as stated in the instructions?

I've just bought a TP-Link AC 1200 Archer A5 v6.0.
It's the same model as the TP-Link Archer C50 v6 (CA/EU/RU).

I can confirm, that OpenWrt version 24.10.0-rc4 works like a charm.
I've just setup everything a few hours ago but ethernet and wifi work just fine, no reboots.
The router acts as a wireguard client which also works just fine.
Thank you very much!

I've used the latest original firmware for boot extraction:

https://www.tp-link.com/de/support/download/archer-a5/v6/#Firmware
Archer A5(EU)_V6_240816 
Archer_A5v6_EU_0.9.1_0.5_up_boot[240816-rel51056]_2024-08-16_15.21.06.bin

jschwartzenberg commented:
@martamusikmaschine that's great news!! Thanks a lot for sharing! Maybe you could create a PR to extend the model name?

martamusikmaschine commented:
Sorry, this was my first time ever doing anything with OpenWrt. I"ve just bought any device because I needed a router asap and tried it out. I have no idea, where to change anything in this project.

rozwell commented:
Naturally, I meant the normal boot or even regular reboot from luci.
In my case, after powering on it takes around ~30s for the power LED to start flashing, which it does for another ~1.5min and then additional 30s to finally boot. Around 2.5~3min seems a little too much for me and that blinking power LED reminds me of flashing the firmware, so I decided to ask before connecting to serial port to see what is actually going on...

martamusikmaschine commented:
Sorry, no idea how long it normally takes. This is the only time I've used OpenWRT. I think 3 minutes is okayish, considering the low computing power of these devices. My normal router also takes over 2 minutes to boot up. But I have no idea, what the blinking means.
